docs/architecture/adr-020-block-size.md

```diff
@@ -0,0 +1,65 @@
+# ADR 020: Limiting txs size inside a block
+
+## Changelog
+
+13-08-2018: Initial Draft
+
+## Context
+
+We currently use MaxTxs to reap txs from the mempool when proposing a block,
+but enforce MaxBytes when unmarshalling a block, so we could easily propose a
+block thats too large to be valid.
+
+We should just remove MaxTxs all together and stick with MaxBytes, and have a
+`mempool.ReapMaxBytes`.
+
+But we can't just reap BlockSize.MaxBytes, since MaxBytes is for the entire block,
+not for the txs inside the block. There's extra amino overhead + the actual
+headers on top of the actual transactions + evidence + last commit.
+
+## Proposed solution
+
+Therefore, we should
+
+1) Get rid of MaxTxs.
+2) Use MaxTxsBytes (rename to MaxDataBytes?) to limit the overall transactions size (Data).
+3) Introduce Max\* variables for header, evidence and last commit:
+  * MaxHeaderBytes
+  * MaxEvidenceBytes
+  * MaxLastCommitBytes
+4) Use these variables to limit the appropriate parts (header, evidence, ...)
+https://github.com/tendermint/tendermint/blob/8a1a79257e8e8b309cd35bb1fe40bf9b3330fd7d/types/block.go#L195
+
+Proposed default values are:
+
+- MaxHeaderBytes - 512 bytes (~200 bytes hashes + 200 bytes - 50 UTF-8 encoded symbols of chain ID)
```

Comment:
This is probably my lack of familiarity, but why are there 200 bytes of hashes? I thought it would just be one merkelized hash at the root of the header.

Reply:
nah, tendermint header will be identical to ABCI header

Reply:
Your right, thanks for the clarification! (I hadn't read through the ABCI header definition before, this makes sense now)